flight/Modules/Logging/logging.c

```diff
@@ -381,25 +353,135 @@ static void loggingTask(void *parameters)
 				read_sector = loggingData.FileSectorNum;
 			}
 			LoggingStatsSet(&loggingData);
-
+			// fall-through to default case
+		default:
+			//  Makes sure that we are not hogging the processor
+			PIOS_Thread_Sleep(10);
+			// Close the file if necessary
+			if (write_open) {
+				PIOS_STREAMFS_Close(streamfs_id);
+				loggingData.MinFileId = PIOS_STREAMFS_MinFileId(streamfs_id);
+				loggingData.MaxFileId = PIOS_STREAMFS_MaxFileId(streamfs_id);
+				LoggingStatsSet(&loggingData);
+				write_open = false;
+			}
 		}
+	}
+}
 
-		i++;
+/**
+ * Forward data from UAVTalk out the serial port
+ * \param[in] data Data buffer to send
+ * \param[in] length Length of buffer
+ * \return -1 on failure
+ * \return number of bytes transmitted on success
+ */
+static int32_t send_data(uint8_t *data, int32_t length)
+{
+	if( PIOS_COM_SendBuffer(logging_com_id, data, length) < 0)
+		return -1;
+
+	written_bytes += length;
+
+	return length;
+}
+
+
+/**
+ * Function for sorting linked list based on logging period and makes
+ * sure that objects with private queues come first.
+ */
+static int info_cmp(struct UAVOLogInfo *info_1, struct UAVOLogInfo *info_2)
+{
+	if ((info_1->queue == shared_queue) && (info_2->queue != shared_queue)){
+		return true;
+	}
+	if ((info_1->queue != shared_queue) && (info_2->queue == shared_queue)){
+		return false;
 	}
+
+	return info_1->logging_period > info_2->logging_period;
 }
 
+/**
+ * @brief Callback for adding an object to the logging queue
+ * @param ev the event
+ */
+static void obj_updated_callback(UAVObjEvent * ev)
+{
+	struct UAVOLogInfo *info;
+	uint32_t now;
+
+	if (loggingData.Operation != LOGGINGSTATS_OPERATION_LOGGING){
+		// We are not logging, so all events are discarded
+		return;
+	}
+	LL_FOREACH(log_info, info) {
+		if (info->obj == ev->obj){
+			now = PIOS_Thread_Systime();
+			if (info->next_log <= now){
+				// compute the time of the next logging operation
+				while (info->next_log < now) {
```

Comment:
This makes me a little nervous since it's not constant time.  I think it should divide, multiply, and add.  If someone sets navigation things to be logged frequently, and only enables a nav mode 10 minutes into flight, this is going to have to iterate a lot.

Maybe like.

```
/* How much do we need to advance the due date so the log would be newly due in this interval. */
info->next_log += ((now - info->next_log) / info->logging_period) * info->logging_period;
/* And advance into the next interval */
info->next_log += info->logging_period;
```